lib/tools/system-calls.js

```diff
@@ -771,41 +772,52 @@ systemCallMethods.waitForDevice = async function (appDeviceReadyTimeout = 30) {
   });
 };
 
+/**
+ * Tries to run privileged shell command without root, otherwise elevates privileges and run them again
+ *
+ * @param {args} Array<string> - Shell arguments commands
+ * @return {boolean} True if the device was switched to root.
+ */
+systemCallMethods.runPrivilegedShell = async function (args) {
+  let isRoot = false;
+  try {
+    await this.shell(args);
+  } catch (err) {
+    if (!err.message.includes('must be root')) {
+      throw err;
+    }
+    // this device needs adb to be running as root to run the specific command.
+    // so try to restart the daemon
+    log.debug(`Device requires adb to be running as root in order to run "adb shell ${quote(args)}". Restarting daemon`);
+    isRoot = await this.root();
+    await this.shell(args);
+  }
+  return isRoot;
+};
+
 /**
  * Reboot the current device and wait until it is completed.
  *
  * @param {number} retries [DEFAULT_ADB_REBOOT_RETRIES] - The maximum number of reboot retries.
  * @throws {Error} If the device failed to reboot and number of retries is exceeded.
  */
 systemCallMethods.reboot = async function (retries = DEFAULT_ADB_REBOOT_RETRIES) {
```

Comment:
any reason we cannot use `reboot` for rebooting the device? 

https://blog.csdn.net/johnnylq/article/details/6401531